### PR TITLE
[TIR] Convert if_then_else intrinsics to if-statements

### PR DESCRIPTION
**Pull request by roastduck**
I added a new pass to convert `tvm_if_then_else` intrinsics to `IfThenElse` statments. For example:

```c++
a[i] = tvm.te.if_then_else(x < y, x, y)
```

is now

```c++
if (x < y) {
    a[i] = x;
} else {
    a[i] = y;
}
```

Rationale:

- `tvm_if_then_else` intrinsics can now enjoy the optimizations of the `LoopPartition` pass after tagged as `likely`. (My condition-heavy sparse application gains up to 2.85x speedup)
- Code lines containing `tvm_if_then_else` is often too long to read. This improves readability.

Changes:

- Add an `IfThenElseIntrinToStmt` pass, which runs just before `LoopPartition`. (**Question 1: Should I run this pass with `simple_mode` on?** Currently no. I don't know if the users like the IR to be more similar to what they write, or the lines to be not too long)
- Added a test for the pass.
- Enable access to the `likely` node in the Python API.
- Fixed the breaking `ThreadSync` pass since it did not allow synchronizing threads in a conditional branch, which may be legal when there are synchronizations in both the then-case and the else-case, like the code piece below. Currently I simply deleted the branch check. (**Question 2: Any better way?**)
```cuda
if (...) {
    __syncthread();
    a[i] = x;
} else {
    __syncthread();
    a[i] = y;
}
```
- Fixed the breaking `LowerWarpMemory` pass by adding an extra branching, which handles the situation where there are only allocates of but no stores into a warp memory.

Algorithm for `IfThenElseIntrinToStmt`:

1. Find a statement.
2. Find the out-most `tvm_if_then_else` intrinsic in it.
3. Put the statement into a new `IfThenElse` statement, and remove the out-most `tvm_if_then_else` intrinsic.
4. Recurse until no more `tvm_if_then_else`.

Requesting reviews and thanks in advance. @tqchen @ZihengJiang @ajtulloch.

### CODE REVIEW

**Comment by roastduck** [comment]
Well, CI for Relay is broken. Closing this PR for now until I fix it.

**Comment by roastduck** [comment]
Fixed. The problem was that some ops consist of too may `tvm_if_then_else`s, such as `stride_set`. I added a threshold to control the max number of `tvm_if_then_else`s in one statement. If the threshold is exceeded, the conversion will not be performed.

Re-requesting reviews. @tqchen @ZihengJiang @ajtulloch.

**Comment by roastduck** [comment]
There are two ways to convert cascading `tvm_if_then_else`s. For example,

```
a = tvm_if_then_else(..., w, x) + tvm_if_then_else(..., y, z)
```

can be converted to be

(1)
```c
if (...) {
    if (...) {
        a = w + y;
    } else {
        a = w + z;
    }
} else {
    if (...) {
        a = x + y;
    } else {
        a = x + z;
    }
}
```

or

(2)
```c
int tmp1
if (...) {
    tmp1 = w;
} else {
    tmp1 = x;
}
int tmp2
if (...) {
    tmp2 = y;
} else {
    tmp2 = z;
}
a = tmp1 + tmp2;
```

The current implementation is (1). Maybe I can change to (2).

**Comment by roastduck** [comment]
Fixed the breaking `fifo_buffer` test by avoiding passing the same buffer as different arguments to a CUDA kernel.

**Comment by tqchen** [comment]
Thanks @roastduck  if the goal is mainly to utilize LoopParition, perhaps a better way would be to insert likely into the condition and enable LoopPartition to deal with if_then_else. Given that it is a quit fundamental  construct that can leads to concise generated code(e.g. ? in CUDA).

**Comment by roastduck** [comment]
> Given that it is a quit fundamental construct that can leads to concise generated code(e.g. ? in CUDA), I think we should keep it throughout the lowering process by default. We can still merge the pass if it is an optional utility

I agree. I will make it default.

> If the goal is mainly to utilize LoopParition, perhaps a better way would be to insert likely into the condition and enable LoopPartition to deal with if_then_else.

I think we are unable to insert `likely` automatically, because we don't know whether the condition is likely or unlikely, so `likely` must be written by users.

Besides, what do you think about the cascading if_then_else issue above?

**Comment by roastduck** [comment]
Made the pass default when lowering.

File: src/tir/transforms/thread_storage_sync.cc [hunk 1]
```diff
@@ -95,8 +95,6 @@ class ThreadSyncPlanner : public StorageAccessVisitor {
         }
       }
       if (sync_before_stmt) {
-        CHECK_EQ(condition_counter(), 0)
-            << "Cannot insert syncs inside condition";
```
Comment (tqchen) [inline src/tir/transforms/thread_storage_sync.cc:99]: This check is necessary for GPUs

**Comment by roastduck** [comment]
I decided not to transform statements that load from or store into shared buffers, so we don't have to worry about the thread sync.

**Comment by roastduck** [comment]
I found that even I disable this transformation when there are loading from or storing into shared buffers, the thread synchronization problem can still occur. I will close this PR until I find a better solution.

**Pull request closed without merging**